ledger/src/blockstore.rs

```diff
@@ -1117,6 +1111,70 @@ impl Blockstore {
             &mut write_batch,
         )?;
 
+        for (erasure_set, working_erasure_meta) in erasure_metas.iter() {
+            if !working_erasure_meta.should_write() {
+                // Not a new erasure meta
+                continue;
+            }
+            let (slot, _) = erasure_set.store_key();
+            // First coding shred from this erasure batch, check the forward merkle root chaining
+            if !self.has_duplicate_shreds_in_slot(slot) {
+                let erasure_meta = working_erasure_meta.as_ref();
+                let shred_id = ShredId::new(
+                    slot,
+                    erasure_meta
+                        .first_received_coding_shred_index()
+                        .expect("First received coding index must exist for all erasure metas"),
+                    ShredType::Code,
+                );
+                let shred = Shred::new_from_serialized_shred(
+                    self.get_shred_from_just_inserted_or_db(&just_inserted_shreds, shred_id)
+                        .expect("Shred indicated by erasure meta must exist")
+                        .into_owned(),
+                )
+                .expect("Shred indicated by erasure meta must be deserializable");
+
+                self.check_forward_chained_merkle_root_consistency(
+                    &shred,
+                    erasure_meta,
+                    &just_inserted_shreds,
+                    &mut merkle_root_metas,
+                    &mut duplicate_shreds,
+                );
+            }
+        }
+
+        for (erasure_set, working_merkle_root_meta) in merkle_root_metas.iter() {
+            if !working_merkle_root_meta.should_write() {
+                // Not a new merkle root meta
+                continue;
+            }
+            let (slot, _) = erasure_set.store_key();
+            // First shred from this erasure batch, check the backwards merkle root chaining
+            if !self.has_duplicate_shreds_in_slot(slot) {
+                let merkle_root_meta = working_merkle_root_meta.as_ref();
+                let shred_id = ShredId::new(
+                    slot,
+                    merkle_root_meta.first_received_shred_index(),
+                    merkle_root_meta.first_received_shred_type(),
+                );
+                let shred = Shred::new_from_serialized_shred(
+                    self.get_shred_from_just_inserted_or_db(&just_inserted_shreds, shred_id)
```

Comment:
similarly here, shouldn't we expect shred to be already in `just_inserted_shreds`?